### ISSUE TITLE
core dumps not generated for gateway envoy

### ISSUE THREAD

**Issue by ostromart** [istio/istio#5745]
proxy_init is not run for envoy on gateway. Most likely because of this, something is missing in the setup such that core dumps don't appear. 
- ulimit -c shows unlimited as it should
- core_patterns looks correct (/etc/istio etc)
I was not able to cause core dumps by sending SIGABRT to any process.
Also tried setting core_pattern to some other dir (/tmp) which required container to run in --privileged but that did not solve the problem. It's not likely to be a perms issue because envoy anyway runs as root.

**Comment by ldemailly** [comment]
Similar/dup of #3064

It was never verified as far as I know

**Comment by ostromart** [comment]
Going through the checklist at https://linux.die.net/man/5/core everything looks good. 
I had a look also to see if the kernel has the option enabled (from /proc.config.gz) and it is:
#
# Executable file formats / Emulations
#
CONFIG_BINFMT_ELF=y
CONFIG_COMPAT_BINFMT_ELF=y
CONFIG_CORE_DUMP_DEFAULT_ELF_HEADERS=y
CONFIG_BINFMT_SCRIPT=y
# CONFIG_HAVE_AOUT is not set
CONFIG_BINFMT_MISC=y
CONFIG_COREDUMP=y

**Comment by ostromart** [comment]
Also tried another experiment: created a new shell process just looping on a sleep. 
Sending SIGABRT to this process _does_ produce a core dump. So this points to envoy running with PR_SET_DUMPABLE set to 0 for some reason.

**Comment by ldemailly** [comment]
try kill -SEGV btw - or even make an actual crash (like deref null ptr)

**Comment by ostromart** [comment]
yeah tried that too. SIGABRT does create a core dump for my test process, but not envoy. The only thing I can think of is that somehow coredump capability is unset somehow in the process of launching envoy. Playing around go syscall after the exec in pilot agent to set this cap to 1.

**Comment by ostromart** [comment]
incidentally, same behavior on ingress, so this is not because proxy_init doesn't run.

**Comment by ostromart** [comment]
The root cause is https://github.com/istio/istio/commit/0759c8572119b8c78eeed1e0bbd3c58efc748092
This PR added a chgrp to the dockerfile. The running envoy process has gid 0, so there's a setgid which has the side effect of disabling PR_DUMPABLE per http://man7.org/linux/man-pages/man2/prctl.2.html.

**Comment by ldemailly** [comment]
is there a workaround/setting to turn it back?

once you do can you make sure to have at least 1 circleci and prow test PR showing the core as downloadable artifact ? tia

**Comment by costinm** [comment]
Fixing this should also unblock   https://github.com/istio/istio/issues/5477 

We should fix this for the release - you can leave the bug open until a test is available, or split that to a separate issue. We also need to integrate it with the new dumpsys script. 

On how to fix it: whatever is safest. I don't know if partially reverting tproxy is worth it - but we can do that if needed. We can also run a command from the agent, or use a script to start, where we can do additional setup.

For 1.0 we'll also want to drop privileges for envoy in tproxy mode - the tproxy is no where near stable.

**Comment by costinm** [comment]
BTW - can we change the pattern from /etc/istio/proxy/core... to /tmp or /var something ? 
I'm pretty sure /etc is not the proper dir for core dumps....

**Comment by ostromart** [comment]
Yes, not sure why this was changed to /etc because it used to be /tmp at some stage. 
Let me explain in more detail why I think there's no easy workaround. 
There's a rather obscure behavior of coredump: if gid is changed, the capability to coredump (controller by a dynamic flag bound to the process) is disabled. See link above for details. This is what appears to be happening here, since removing the chrgp in the dockerfille restores coredumps. 
Now the reason the setgid happens (I believe) is because envoy itself calls it. It certainly doesn't happen in the proxy agent code, since that only does a straightforward exec. And if you look at the gid of the running process, it is 0, not the one in chgrp (1ee7). So we cannot change this easily unless we change envoy code (and there are probably good reasons for envoy to do the setgid).

**Comment by ostromart** [comment]
Plan for now is to remove the chgrp, which will temporarily break tproxy. 
To restore tproxy in next release, we have some options (in rough order of complexity):
1. Bifurcate dockerfiles into regular/tproxy
2. Add prctl() syscall in envoy which will restore ability to core dump (there could be security objections to this).
3. Move the chgrp into pilot agent and set it there conditionally (based on tproxy support)
4. Figure out some way to make tproxy work without the chgrp.
